### PR TITLE
Fix segfault caused by pop from numba.typed.List

### LINKED ISSUE #5264: Heap corruption of Lists of numpy arrays
Lists of numpy arrays (at least of float 1D type) cannot be brought to length zero without heap corruption. Best guess at this point: refcount issue and double free of allocated memory.

Also, Lists of numpy scalars for some reason do not suffer this issue.

Minimal reproducer:

```python
import numpy as np
from numba import  float64
from numba.typed import List

float_array = float64[:]

l = List.empty_list(float_array)
l.append(np.ones(3,dtype=np.float64))
l.pop()
````

### PR DESCRIPTION
**Pull request by esc**
Fixes #5264

### CODE REVIEW

**Comment by esc** [comment]
May also fix #4960

File: numba/cext/listobject.c [hunk 1]
```diff
@@ -441,6 +399,50 @@ numba_list_resize(NB_List *lp, Py_ssize_t newsize) {
...
+/* Delete a single item.
+ *
+ * lp: a list
+ * index: the index of the item to delete
+ *
+ * */
+int
+numba_list_delitem(NB_List *lp, Py_ssize_t index) {
+    // FIXME the following line is the lazy implementation
+    //result = numba_list_delete_slice(lp, index, index+1, 1);
```
Comment (stuartarchibald) [inline numba/cext/listobject.c:411]: Remove?

File: numba/cext/listobject.c [hunk 2]
```diff
@@ -441,6 +399,50 @@ numba_list_resize(NB_List *lp, Py_ssize_t newsize) {
...
+    // check index is valid
+    // FIXME: this can be (and probably is) checked at the compiler level
+    if (!valid_index(index, lp->size)) {
+        return LIST_ERR_INDEX;
+    }
+    // obtain item and decref if needed
+    loc = lp->items + lp->item_size * index;
+    list_decref_item(lp, loc);
+    if (index != lp->size - 1) {
+        // delitem from somewhere other than the end, incur the dreaded memory copy
```
Comment (stuartarchibald) [inline numba/cext/listobject.c:428]:
```suggestion
        // delitem from somewhere other than the end, incur the memory copy
```

File: numba/cext/listobject.c [hunk 3]
```diff
@@ -441,6 +399,50 @@ numba_list_resize(NB_List *lp, Py_ssize_t newsize) {
...
 }
 
+/* Delete a single item.
+ *
+ * lp: a list
+ * index: the index of the item to delete
+ *
+ * */
+int
+numba_list_delitem(NB_List *lp, Py_ssize_t index) {
```
Comment (stuartarchibald) [inline numba/cext/listobject.c:409]: The index is restricted to C semantics here, i.e. non-negative? If so perhaps add to docstring.

File: numba/cext/listobject.c (hunk 1 above)
Comment (esc) [reply numba/cext/listobject.c:411]: done

File: numba/cext/listobject.c (hunk 3 above)
Comment (esc) [reply numba/cext/listobject.c:409]: The `index` appears in many of the functions, so for consistency it should probably be set everywhere. How about something like:

```
 * index: the index of the item to delete (must be in range 0 <= index < len(list))
```

File: numba/cext/listobject.c (hunk 2 above)
Comment (esc) [reply numba/cext/listobject.c:428]: done

**Comment by esc** [comment]
@stuartarchibald I have addressed all comments except one and added the reproducers. Let me know how to proceed.

File: numba/cext/listobject.c (hunk 3 above)
Comment (stuartarchibald) [reply numba/cext/listobject.c:409]: Sounds good.

**Comment by stuartarchibald** [comment]
> @stuartarchibald I have addressed all comments except one and added the reproducers. Let me know how to proceed.

Thanks, have responded to the remaining item.

File: numba/cext/listobject.c (hunk 3 above)
Comment (esc) [reply numba/cext/listobject.c:410]: updated!

**Pull request merged**